src/rootfs/udevd/pkg/drivers/scsi/scsi.go

```diff
@@ -0,0 +1,392 @@
+// pointer used for either reading or writing.
+type GenericIOHeader struct {
+	InterfaceID    int32   // [i] 'S' (required)
+	DxferDirection int32   // [i]
+	CmdLen         uint8   // [i]
+	MxSbLen        uint8   // [i]
+	IovecCount     uint16  // [i]
+	DxferLen       uint32  // [i]
+	Dxferp         *byte   // [i], [*io]
+	Cmdp           *uint8  // [i], [*i]
+	Sbp            *uint8  // [i], [*o]
+	Timeout        uint32  // [i] unit: millisecs
+	Flags          uint32  // [i]
+	PackID         int32   // [i->o]
+	pad0           [4]byte // nolint: structcheck
+	UsrPtr         *byte   // [i->o]
+	Status         uint8   // [o]
+	MaskedStatus   uint8   // [o]
+	MsgStatus      uint8   // [o]
+	SbLenWr        uint8   // [o]
+	HostStatus     uint16  // [o]
+	DriverStatus   uint16  // [o]
+	Resid          int32   // [o]
+	Duration       uint32  // [o]
+	Info           uint32  // [o]
+}
+
+// ID represents the set of SCSI device identifiers.
+type ID struct {
+	Vendor             string
+	Model              string
+	Revision           string
+	Type               Type
+	Kernel             string
+	Serial             string
+	SerialShort        string
+	UseSg              int32
+	UnitSerialNumber   string
+	WWN                string
+	WWNVendorExtension string
+	TgptGroup          string
+}
+
+// Device represents a SCSI device.
+type Device struct {
+	*ID
+
+	f *os.File
+}
+
+func open(name string) (f *os.File, err error) {
+	return os.OpenFile(name, os.O_RDONLY|unix.O_NONBLOCK|unix.O_CLOEXEC, 0)
+}
+
+// NewDevice returns a SCSI device.
+// See https://www.tldp.org/HOWTO/SCSI-Generic-HOWTO/iddriver.html.
+func NewDevice(name string) (device *Device, err error) {
+	var version uint32
+	ptr := uintptr(unsafe.Pointer(&version))
+
+	var f *os.File
+	if f, err = open(name); err != nil {
+		return nil, err
+	}
+
+	if err = ioctl(f.Fd(), SgGetVersionNum, ptr); err != nil {
+		return nil, err
+	}
+
+	if version < 30000 {
+		return nil, errors.Errorf("no sg driver found for device at %s", name)
+	}
+
+	device = &Device{
+		f: f,
+		ID: &ID{
+			UseSg: 3,
+		},
+	}
+
+	return device, nil
+}
+
+// Close closes the device.
+func (dvc *Device) Close() error {
+	return dvc.f.Close()
+}
+
+// Identify queries a SCSI device via the SCSI INQUIRY vital product data (VPD)
+// page 0x80 and 0x83.
+func (dvc *Device) Identify() (err error) {
+	// Standard inquiry.
+	if err = dvc.StandardInquiry(); err != nil {
+		return errors.Errorf("standard inquiry failed: %v", err)
+	}
+
+	// Page 0 inquiry.
+	var pages []byte
+	if pages, err = dvc.Page0Inquiry(); err != nil {
+		return errors.Errorf("page 0 inquiry failed: %v", err)
+	}
+	// TODO(andrewrynhard): Prevent Page80Inquiry from running twice.
+	for _, page := range pages {
+		if page == Page80 {
+			// Page 80 inquiry.
+			if err = dvc.Page80Inquiry(); err != nil {
+				return errors.Errorf("page 80 inquiry failed: %v", err)
+			}
+		}
+		if page == Page83 {
```

Comment:
is it worth combining these conditionals to fix TODO?

```
		if ( page == Page80 || page == Page83 ) {
			// Page 80 inquiry.
			if err = dvc.Page80Inquiry(); err != nil {
				return errors.Errorf("page 80 inquiry failed: %v", err)
			}
		}

		if page == Page83 {
			// Page 83 inquiry.
			if err = dvc.Page83Inquiry(); err != nil {
				return errors.Errorf("page 83 inquiry failed: %v", err)
			}
		}
```

Reply:
Perfect. Good idea.